### PR TITLE
New: implement eslint/rfcs#63

### PR DESCRIPTION
**Pull request by wmhilton**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[x] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This is _heavily_ based on https://github.com/eslint/eslint/pull/11487 but as per the RFC the names of the options have been changed.

I also rebased it and tried to get all the tests passing. There's still one failing locally. Not sure what that's about yet.
image

#### Is there anything you'd like reviewers to focus on?

My original PR was a one-liner. And now my PR is largely someone elses code 😬 so aaaaany help would be appreciated. I don't care about credit, I just want this feature shipped so my team stops burning through CircleCI minutes waiting to lint files that don't need to be linted.

#### background

Currently CI users resort to hacks such as manually modifying the `mtime` of all their files:
https://github.com/salto-io/salto/pull/395/files#diff-5d8151d7fb61e762279563e3d38abd41R1-R17

Previous attempts to address this:
- https://github.com/eslint/eslint/issues/11319
- https://github.com/eslint/eslint/pull/11487
- https://github.com/eslint/eslint/issues/11490
- https://github.com/royriojas/file-entry-cache/pull/14

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks for attempting this. I’m not comfortable with reading an environment variable for this purpose — that’s just one assumption too many when determining how ESLint should behave.

We did recently merge this RFC that covers this use case: https://github.com/eslint/rfcs/pull/63

**Comment by wmhilton** [comment]
Oh that's fantastic! That's great. Is there already an implementation of RFC 63? If not, I can pivot to implementing it.

**Comment by wmhilton** [comment]
@c-home anything I can do to help?

**Comment by nzakas** [comment]
I don't believe implementation has started yet, so you can feel free!

**Comment by c-home** [comment]
> @c-home anything I can do to help?

Hi, I started implementing the RFC but haven't gotten a chance to add testing yet.
If you would like to implement it that is great :)

File: docs/user-guide/command-line-interface.md [hunk 1]
```diff
@@ -440,6 +441,16 @@ Example:
 
     eslint "src/**/*.js" --cache --cache-location "/Users/user/.eslintcache/"
 
+#### `--cache-strategy`
+
+Strategy to use for detecting changed files. Can be either `metadata` or `content`. If no strategy is specified, `metadata` will be used.
```
Comment (nzakas) [inline docs/user-guide/command-line-interface.md:446]:
```suggestion
Strategy for the cache to use for detecting changed files. Can be either `metadata` or `content`. If no strategy is specified, `metadata` will be used.
```

File: lib/eslint/eslint.js [hunk 2]
```diff
@@ -43,6 +43,7 @@ const { version } = require("../../package.json");
  * @property {ConfigData} [baseConfig] Base config object, extended by all configs used with this instance
  * @property {boolean} [cache] Enable result caching.
  * @property {string} [cacheLocation] The cache file to use instead of .eslintcache.
+ * @property {"metadata" | "content"} [cacheStrategy] The strategy used to detect changed files.
```
Comment (nzakas) [inline lib/eslint/eslint.js:46]: This file shouldn't have any Node.js-specific information inside of it. It doesn't directly access the cache, so I think we can safely remove all of the cache-related changes from this file.

File: lib/options.js [hunk 3]
```diff
@@ -214,6 +214,14 @@ module.exports = optionator({
             type: "path::String",
             description: "Path to the cache file or directory"
         },
+        {
+            option: "cache-strategy",
+            dependsOn: ["cache"],
+            type: "String",
+            default: "metadata",
+            enum: ["metadata", "content"],
+            description: "Strategy to use for detecting changed files"
```
Comment (nzakas) [inline lib/options.js:223]:
```suggestion
            description: "Strategy to use for detecting changed files in the cache"
```

**Review by nzakas — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing, this looks really good. Just a few small things to clean up and I'd like to get a few other sets of eyes on this, too.

**Comment by nzakas** [comment]
@c-home if you have the time to take a look at this implementation, we'd appreciate your feedback.

**Comment by c-home** [comment]
> @c-home if you have the time to take a look at this implementation, we'd appreciate your feedback.

Had a quick look and it looks great. Will have a closer look soon. Thank you for implementing this wmhilton.

File: lib/cli-engine/lint-result-cache.js [hunk 4]
```diff
@@ -76,17 +115,28 @@ class LintResultCache {
          *    was previously linted
          * If any of these are not true, we will not reuse the lint results.
          */
-
         const fileDescriptor = this.fileEntryCache.getFileDescriptor(filePath);
         const hashOfConfig = hashOfConfigFor(config);
-        const changed = fileDescriptor.changed || fileDescriptor.meta.hashOfConfig !== hashOfConfig;
+        const changed =
+            fileDescriptor.changed ||
+            fileDescriptor.meta.hashOfConfig !== hashOfConfig;
```
Comment (c-home) [inline lib/cli-engine/lint-result-cache.js:122]: Will `changed` be `true` if the `cacheStrategy` changes? Related to [this issue](https://github.com/eslint/rfcs/pull/63#issuecomment-676463622) raised in the rfc.

**Comment by c-home** [comment]
For users of eslint cache, this is something to consider - https://github.com/eslint/rfcs/pull/68.

File: lib/eslint/eslint.js (hunk 2 above)
Comment (nzakas) [reply lib/eslint/eslint.js:46]: Oops, please ignore this comment. I misread the filename. This looks good. :+1:

**Comment by wmhilton** [comment]
Ugh. As pointed out in [this comment](https://github.com/eslint/rfcs/pull/68#issuecomment-707414470) some very notable rulesets such as `@typescript-eslint` are incompatible with `--cache` by design. That really puts a damper on my enthusiasm for this, since anything less than 100% reproducible isn't good for our CI use case at Stoplight. :-(

**Comment by nzakas** [comment]
It seems like you are pretty close to finishing this, and this is something we plan on doing regardless. Do you want to finish it up so we can merge it?

We do still need to solve for the case where plugins are doing unexpected things, but there are plenty of folks who would benefit from this PR in the meantime.

**Comment by royriojas** [comment]
hi @nzakas @wmhilton I've published a new version of `file-entry-cache`. Version `6.0.0`. This version has no new functionality but updates the dependencies to latest to remove security warnings.

It is a major upgrade to avoid any potential issues since I've aligned the supported node versions with the ones `eslint` is currently using.

**Comment by nzakas** [comment]
@royriojas thanks for the heads up. Can you open an issue so that info doesn’t get lost here?

**Comment by vaniyokk** [comment]
Any update on this?

**Comment by nzakas** [comment]
@vaniyokk it appears this has been abandoned. Do you want to take it up?

**Comment by kirill-konshin** [comment]
> @vaniyokk it appears this has been abandoned. Do you want to take it up?

This is quite sad. In our case on CI with repo having around 10K modules ESLint takes waaaay too much time, since the cache is not used at all between pushes (due to timestamps inconsistency, Git does not maintain them, and GitLab does not have workspaces so far).

**Comment by royriojas** [comment]
The merge request seems to be almost ready, isn't it? @nzakas. I think it is only a matter of fix the tests

**Comment by snitin315** [comment]
Can I work on this and finish the PR ???

**Comment by nzakas** [comment]
@royriojas yes, it looks basically done at this point.

@snitin315 yes, you can definitely take it over. You should be able to pull the branch from where this PR was created and then create a new PR with that code. When you do, just comment here with the new PR number so we can close this and move our focus over to the new PR. Thanks!

**Comment by mdjermanovic** [comment]
Closing as there's a new PR #14119

**Pull request closed without merging**